metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py

```diff
@@ -0,0 +1,1096 @@
+from datahub.ingestion.glossary.classification_mixin import (
+    SAMPLE_SIZE_MULTIPLIER,
+    ClassificationHandler,
+    classification_workunit_processor,
+)
+from datahub.ingestion.source.bigquery_v2.bigquery_audit import (
+    BigqueryTableIdentifier,
+    BigQueryTableRef,
+)
+from datahub.ingestion.source.bigquery_v2.bigquery_config import BigQueryV2Config
+from datahub.ingestion.source.bigquery_v2.bigquery_data_reader import BigQueryDataReader
+from datahub.ingestion.source.bigquery_v2.bigquery_helper import (
+    unquote_and_decode_unicode_escape_seq,
+)
+from datahub.ingestion.source.bigquery_v2.bigquery_report import BigQueryV2Report
+from datahub.ingestion.source.bigquery_v2.bigquery_schema import (
+    SCHEMA_PARALLELISM,
+    BigqueryColumn,
+    BigqueryDataset,
+    BigqueryProject,
+    BigQuerySchemaApi,
+    BigqueryTable,
+    BigqueryTableSnapshot,
+    BigqueryView,
+)
+from datahub.ingestion.source.bigquery_v2.common import (
+    BQ_EXTERNAL_DATASET_URL_TEMPLATE,
+    BQ_EXTERNAL_TABLE_URL_TEMPLATE,
+)
+from datahub.ingestion.source.bigquery_v2.profiler import BigqueryProfiler
+from datahub.ingestion.source.common.subtypes import (
+    DatasetContainerSubTypes,
+    DatasetSubTypes,
+)
+from datahub.ingestion.source.sql.sql_utils import (
+    add_table_to_schema_container,
+    gen_database_container,
+    gen_schema_container,
+    get_domain_wu,
+)
+from datahub.ingestion.source_report.ingestion_stage import (
+    METADATA_EXTRACTION,
+    PROFILING,
+)
+from datahub.metadata.com.linkedin.pegasus2avro.common import (
+    Status,
+    SubTypes,
+    TimeStamp,
+)
+from datahub.metadata.com.linkedin.pegasus2avro.dataset import (
+    DatasetProperties,
+    ViewProperties,
+)
+from datahub.metadata.com.linkedin.pegasus2avro.schema import (
+    ArrayType,
+    BooleanType,
+    BytesType,
+    DateType,
+    MySqlDDL,
+    NullType,
+    NumberType,
+    RecordType,
+    SchemaField,
+    SchemaFieldDataType,
+    SchemaMetadata,
+    StringType,
+    TimeType,
+)
+from datahub.metadata.schema_classes import (
+    DataPlatformInstanceClass,
+    GlobalTagsClass,
+    TagAssociationClass,
+)
+from datahub.sql_parsing.schema_resolver import SchemaResolver
+from datahub.utilities.file_backed_collections import FileBackedDict
+from datahub.utilities.hive_schema_to_avro import (
+    HiveColumnToAvroConverter,
+    get_schema_fields_for_hive_column,
+)
+from datahub.utilities.mapping import Constants
+from datahub.utilities.perf_timer import PerfTimer
+from datahub.utilities.ratelimiter import RateLimiter
+from datahub.utilities.registries.domain_registry import DomainRegistry
+
+logger: logging.Logger = logging.getLogger(__name__)
+# Handle table snapshots
+# See https://cloud.google.com/bigquery/docs/table-snapshots-intro.
+SNAPSHOT_TABLE_REGEX = re.compile(r"^(.+)@(\d{13})$")
+CLUSTERING_COLUMN_TAG = "CLUSTERING_COLUMN"
+
+
+class BigQueryProjectSchemaGenerator:
```

Comment:
is this specific to a single project?

Reply:
Renaming this to `BigQuerySchemaGenerator`. Currently calling its `get_project_workunits` for every project. It maintains the cache of entity references across all projects that it has processed.
```
       self.bq_schema_extractor.view_refs_by_project,
        self.bq_schema_extractor.view_definitions,
        self.bq_schema_extractor.snapshot_refs_by_project,
        self.bq_schema_extractor.snapshots_by_ref,
        self.bq_schema_extractor.table_refs,
```